quicklens/src/test/scala-3/com/softwaremill/quicklens/test/ModifyAndTypeTest.scala

```diff
@@ -0,0 +1,88 @@
+package com.softwaremill.quicklens
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import ModifyAndTypeTest._
+
+object ModifyAndTypeTest {
+  case class A(a: Int) extends B
+  trait B {
+    def a: Int
+  }
+
+  case class A1(a: Int)
+  
+  sealed trait T
+  case class C(a: Int) extends T with B
+
+  sealed trait T1
+  case class C1(a: Int) extends T1
+}
+
+class ModifyAndTypeTest extends AnyFlatSpec with Matchers {
+  it should "modify an & type object" in {
+    val ab: A & B = A(0)
+
+    val modified = ab.modify(_.a).setTo(1)
+
+    modified.a shouldBe 1
+  }
+
+  it should "modify an & type object 1" in {
+    val ab: B & A = A(0)
+
+    val modified = ab.modify(_.a).setTo(1)
+
+    modified.a shouldBe 1
+  }
+
+  it should "modify an & type object 2" in {
+    val ab: B & A1 = new A1(0) with B
+
+    val modified = ab.modify(_.a).setTo(1)
+
+    modified.a shouldBe 1
+  }
+
+  it should "modify an & type object 3" in {
+    val ab: A1 & B = new A1(0) with B
+
+    val modified = ab.modify(_.a).setTo(1)
+
+    modified.a shouldBe 1
+  }
+
+  // TODO this is an implemenation limitation for now, since anonymous classes crash on runtime
```

Comment:
limitation of quicklens, Scala 3? If the latter, is there a dotty ticket maybe?

Reply:
> limitation of quicklens, Scala 3?

`quicklens`. I think.

I'm not sure if there should be a dotty ticket for it since we're using `asInstanceOf`.
Also, the problem occurs only for the last two tests - the ones with anonymous classes.
It's not hard to minimize, because we can just type the generated code by hand:
```scala
sealed trait T
case class A(a: Int) extends T
trait B {
  def a: Int
}

val ab: T & B = new A(0) with B

val res: T & B = // if we change the ascription to `A & B` it works
  if(ab.isInstanceOf[A]) then {
    ab.asInstanceOf[A].copy().asInstanceOf[A & B]
  } else {
    ???
  }
```
Scastie: https://scastie.scala-lang.org/KacperFKorban/s7A3euoTST2AoPYcTug8iA/5

Maybe there is some other way to write it so that it won't crash on runtime :thinking:

The only weird thing is that it doesn't crash if we change the type ascription to the more specific one.

Reply:
Note: when I rewrite this using `with` instead of  ¨&` and try it with Scala 2.13, it works: https://scastie.scala-lang.org/0vAFV0B8SDmZrjvbRj3jsA

The same code with `with` still does not work in Scala 3.

To me this looks like a Scala 3 bug.

Reply:
Reported in https://github.com/lampepfl/dotty/issues/15952

Reply:
Ok so now that we know it's a quicklens not a dotty issue, can we do anything about it? :)

Reply:
I don't see an immediate solution. The problem is in the generated bytecode, so no matter what we do `ab.asInstanceOf[A].copy()` will have to be cast to `T & B`. Which is impossible.

Also if we want to have a discussion about it, maybe a new issue will be a better place for it :smiley:

Reply:
Sure :) https://github.com/softwaremill/quicklens/issues/100